File: windows/TokenTracker.Windows.Core/UsageForecast.cs
namespace TokenTracker.Windows.Core;

/// Which usage window a forecast is computed against.
public enum ForecastWindow
{
    FiveHour,
    SevenDay
}

/// A depletion forecast for one provider/window: how fast remaining budget is
/// being consumed and when, at that pace, it reaches zero. Derived purely from
/// locally stored history — it makes no network call. Mirrors the Swift
/// <c>UsageForecast</c> in <c>Sources/TokenTrackerCore/UsageForecast.swift</c>.
public sealed record UsageForecast(
    double BurnPerHour,
    double SecondsToEmpty,
    DateTimeOffset EmptyAt,
    bool WillEmptyBeforeReset);

public static class UsageForecaster
{
    /// Minimum observed span before a forecast is produced, so a couple of
    /// near-simultaneous samples can't extrapolate to a wild ETA.
    public static readonly TimeSpan MinimumSpan = TimeSpan.FromMinutes(10);

    /// Newest sample must be at most this old, so a window that stopped
    /// reporting (or a long pause) doesn't keep projecting from stale data.
    public static readonly TimeSpan MaximumSampleAge = TimeSpan.FromMinutes(30);

    /// Project when the given provider/window runs out at its recent burn rate.
    /// Returns <c>null</c> when there isn't enough signal: fewer than two
    /// post-reset samples, a span under <see cref="MinimumSpan"/>, a newest
    /// sample older than <see cref="MaximumSampleAge"/>, or a
    /// flat/replenishing window.
    public static UsageForecast? Forecast(
        IReadOnlyList<UsageHistoryEntry> entries,
        Provider provider,
        ForecastWindow window,
        DateTimeOffset? resetAt,
        DateTimeOffset? now = null)
    {
        var current = now ?? DateTimeOffset.Now;

        var points = entries
            .Select(entry => (T: entry.RecordedAt, R: Remaining(entry.Snapshot.Usage(provider), window)))
            .Where(point => point.R is not null)
            .Select(point => (point.T, R: point.R!.Value))
            .OrderBy(point => point.T)
            .ToList();

        if (points.Count < 2)
        {
            return null;
        }

        // Trim to the current window instance: drop everything up to and
        // including the last upward jump (a reset refilling the budget),
        // otherwise a reset reads as negative consumption.
        var startIndex = 0;
        for (var i = 1; i < points.Count; i++)
        {
            if (points[i].R > points[i - 1].R)
            {
                startIndex = i;
            }
        }

        var segment = points.Skip(startIndex).ToList();
        if (segment.Count < 2)
        {
            return null;
        }

        var first = segment[0];
        var last = segment[^1];

        var elapsed = (last.T - first.T).TotalSeconds;
        if (elapsed < MinimumSpan.TotalSeconds)
        {
            return null;
        }

        if ((current - last.T).TotalSeconds > MaximumSampleAge.TotalSeconds)
        {
            return null;
        }

        double drop = first.R - last.R;
        if (drop <= 0)
        {
            return null; // steady or replenishing → no forecast
        }

        var burnPerHour = drop / (elapsed / 3600.0);
        if (burnPerHour <= 0)
        {
            return null;
        }

        var secondsToEmpty = last.R / burnPerHour * 3600.0;
        var emptyAt = current.AddSeconds(secondsToEmpty);
        var willEmptyBeforeReset = resetAt.HasValue && emptyAt < resetAt.Value;

        return new UsageForecast(burnPerHour, secondsToEmpty, emptyAt, willEmptyBeforeReset);
    }

    /// Compact, language-neutral duration like <c>2h 10m</c> / <c>45m</c> / <c>&lt;1m</c>.
    public static string DurationText(double seconds)
    {
        var totalMinutes = (int)(seconds / 60);
        if (totalMinutes <= 0)
        {
            return "<1m";
        }

        if (totalMinutes < 60)
        {
            return $"{totalMinutes}m";
        }

        return $"{totalMinutes / 60}h {totalMinutes % 60}m";
    }

    private static int? Remaining(ProviderUsage usage, ForecastWindow window) =>
        window == ForecastWindow.FiveHour ? usage.RemainingPercent5h : usage.RemainingPercent7d;
}

public static class UsageForecastText
{
    /// The per-provider menu line, or <c>null</c> when there is no forecast to show.
    public static string? MenuLine(UsageForecast? forecast, Localizer localizer)
    {
        if (forecast is null)
        {
            return null;
        }

        var line = $"{localizer.Text(L10nKey.ForecastLabel)}: ~{UsageForecaster.DurationText(forecast.SecondsToEmpty)}";
        if (forecast.WillEmptyBeforeReset)
        {
            line += $" · {localizer.Text(L10nKey.ForecastBeforeReset)}";
        }

        return line;
    }
}

/// One provider/window forecast fed to the predictive-alert evaluator.
public sealed record ForecastAlertInput(
    Provider Provider,
    ForecastWindow Window,
    UsageForecast Forecast,
    DateTimeOffset? ResetAt);

public static class UsageForecastAlert
{
    /// Emit one alert per input whose budget is projected to empty before its
    /// reset. <paramref name="enabled"/> is the caller's combined gate
    /// (notifications on and depletion alert on); the id includes the reset
    /// instant so it dedupes per window instance, matching the reset-proximity
    /// alert convention.
    public static IReadOnlyList<UsageAlertCandidate> Candidates(
        IReadOnlyList<ForecastAlertInput> inputs,
        bool enabled,
        Localizer? localizer = null)
    {
        if (!enabled)
        {
            return Array.Empty<UsageAlertCandidate>();
        }

        localizer ??= new Localizer(AppLanguage.English);
        var result = new List<UsageAlertCandidate>();
        foreach (var input in inputs)
        {
            if (!input.Forecast.WillEmptyBeforeReset || input.ResetAt is null)
            {
                continue;
            }

            var windowLabel = input.Window == ForecastWindow.FiveHour ? "5h" : "7d";
            var resetId = input.ResetAt.Value.ToUnixTimeSeconds();
            var name = input.Provider == Provider.Claude ? "Claude" : "Codex";
            result.Add(new UsageAlertCandidate(
                $"{input.Provider.ToId()}-{windowLabel}-empty-before-reset-{resetId}",
                localizer.Text(L10nKey.DepletionAlertTitle),
                $"{name} {windowLabel}: ~{UsageForecaster.DurationText(input.Forecast.SecondsToEmpty)} → 0% (before reset)"));
        }

        return result;
    }
}

File: Sources/TokenTrackerCore/UsageForecast.swift
import Foundation

/// Which usage window a forecast is computed against.
public enum ForecastWindow: Sendable {
    case fiveHour
    case sevenDay
}

/// A depletion forecast for one provider/window: how fast remaining budget is
/// being consumed and when, at that pace, it reaches zero. Derived purely from
/// locally stored history — it makes **no** network call.
public struct UsageForecast: Equatable, Sendable {
    /// Remaining-percentage consumed per hour (always > 0 for a produced forecast).
    public let burnPerHour: Double
    public let secondsToEmpty: Double
    public let emptyAt: Date
    /// True when the projected empty instant falls before the window's reset.
    public let willEmptyBeforeReset: Bool

    public init(burnPerHour: Double, secondsToEmpty: Double, emptyAt: Date, willEmptyBeforeReset: Bool) {
        self.burnPerHour = burnPerHour
        self.secondsToEmpty = secondsToEmpty
        self.emptyAt = emptyAt
        self.willEmptyBeforeReset = willEmptyBeforeReset
    }
}

public enum UsageForecaster {
    /// Minimum observed span before a forecast is produced, so a couple of
    /// near-simultaneous samples can't extrapolate to a wild ETA.
    public static let minimumSpan: TimeInterval = 600 // 10 minutes

    /// Newest sample must be at most this old, so a window that stopped
    /// reporting (or a long pause) doesn't keep projecting from stale data.
    public static let maximumSampleAge: TimeInterval = 1800 // 30 minutes

    /// Project when the given provider/window runs out at its recent burn rate.
    /// Returns `nil` when there isn't enough signal: fewer than two post-reset
    /// samples, a span under `minimumSpan`, a newest sample older than
    /// `maximumSampleAge`, or a flat/replenishing window.
    public static func forecast(
        entries: [UsageHistoryEntry],
        provider: Provider,
        window: ForecastWindow,
        resetAt: Date?,
        now: Date = Date()
    ) -> UsageForecast? {
        let points: [(t: Date, r: Int)] = entries.compactMap { entry in
            guard let remaining = remaining(entry.snapshot.usage(for: provider), window) else {
                return nil
            }
            return (entry.recordedAt, remaining)
        }.sorted { $0.t < $1.t }

        guard points.count >= 2 else { return nil }

        // Trim to the current window instance: drop everything up to and
        // including the last upward jump (a reset refilling the budget),
        // otherwise a reset reads as negative consumption.
        var startIndex = 0
        for index in 1..<points.count where points[index].r > points[index - 1].r {
            startIndex = index
        }
        let segment = Array(points[startIndex...])
        guard segment.count >= 2, let first = segment.first, let last = segment.last else {
            return nil
        }

        let elapsed = last.t.timeIntervalSince(first.t)
        guard elapsed >= minimumSpan else { return nil }
        guard now.timeIntervalSince(last.t) <= maximumSampleAge else { return nil }

        let drop = Double(first.r - last.r)
        guard drop > 0 else { return nil } // steady or replenishing → no forecast

        let burnPerHour = drop / (elapsed / 3600)
        guard burnPerHour > 0 else { return nil }

        let secondsToEmpty = Double(last.r) / burnPerHour * 3600
        let emptyAt = now.addingTimeInterval(secondsToEmpty)
        let willEmptyBeforeReset = resetAt.map { emptyAt < $0 } ?? false

        return UsageForecast(
            burnPerHour: burnPerHour,
            secondsToEmpty: secondsToEmpty,
            emptyAt: emptyAt,
            willEmptyBeforeReset: willEmptyBeforeReset
        )
    }

    /// Compact, language-neutral duration like `2h 10m` / `45m` / `<1m`.
    public static func durationText(_ seconds: Double) -> String {
        let totalMinutes = Int(seconds / 60)
        if totalMinutes <= 0 { return "<1m" }
        if totalMinutes < 60 { return "\(totalMinutes)m" }
        return "\(totalMinutes / 60)h \(totalMinutes % 60)m"
    }

    private static func remaining(_ usage: ProviderUsage, _ window: ForecastWindow) -> Int? {
        switch window {
        case .fiveHour: return usage.remainingPercent5h
        case .sevenDay: return usage.remainingPercent7d
        }
    }
}

public enum UsageForecastText {
    /// The per-provider menu line, or `nil` when there is no forecast to show.
    public static func menuLine(forecast: UsageForecast?, localizer: Localizer) -> String? {
        guard let forecast else { return nil }
        var line = "\(localizer.text(.forecastLabel)): ~\(UsageForecaster.durationText(forecast.secondsToEmpty))"
        if forecast.willEmptyBeforeReset {
            line += " · \(localizer.text(.forecastBeforeReset))"
        }
        return line
    }
}

/// One provider/window forecast fed to the predictive-alert evaluator.
public struct ForecastAlertInput: Sendable {
    public let provider: Provider
    public let window: ForecastWindow
    public let forecast: UsageForecast
    public let resetAt: Date?

    public init(provider: Provider, window: ForecastWindow, forecast: UsageForecast, resetAt: Date?) {
        self.provider = provider
        self.window = window
        self.forecast = forecast
        self.resetAt = resetAt
    }
}

public enum UsageForecastAlert {
    /// Emit one alert per input whose budget is projected to empty before its
    /// reset. `enabled` is the caller's combined gate (notifications on **and**
    /// depletion alert on); the id includes the reset instant so it dedupes per
    /// window instance, matching the reset-proximity alert convention.
    public static func candidates(
        inputs: [ForecastAlertInput],
        enabled: Bool,
        localizer: Localizer = Localizer(language: .english)
    ) -> [UsageAlertCandidate] {
        guard enabled else { return [] }

        return inputs.compactMap { input in
            guard input.forecast.willEmptyBeforeReset, let resetAt = input.resetAt else {
                return nil
            }
            let windowLabel = input.window == .fiveHour ? "5h" : "7d"
            let resetID = Int(resetAt.timeIntervalSince1970)
            return UsageAlertCandidate(
                id: "\(input.provider.rawValue)-\(windowLabel)-empty-before-reset-\(resetID)",
                title: localizer.text(.depletionAlertTitle),
                body: "\(input.provider.displayName) \(windowLabel): ~\(UsageForecaster.durationText(input.forecast.secondsToEmpty)) → 0% (before reset)"
            )
        }
    }
}
